### PR TITLE
feat(remote_wal): add skeleton for remote wal related to datanode

### PR DESCRIPTION
**Pull request by niebayes**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

- Allocated wal options, including kafka topic, will be persisted in `RegionInfo` of `DatanodeTableValue`, and will be passed to datanodes through `RegionCreateRequest`. See: https://github.com/GreptimeTeam/greptimedb/pull/2933
- Upon initializing a region server, each region's wal options are decomposed from the `RegionInfo` in `DatanodeTableValue`. Those wal options are then merged into region options and passed to region workers.
- When a region worker receives a region create/open request, it extracts wal options from region options therein. The wal options are used to construct wal entry namespace when replaying memtable. On the other hand, the wal options are stored in `MitoRegion` to be used during writing to a region.
- When a region worker receives a `RegionWriteRequest`, it constructs wal entry namespaces with wal options. Entries are then written into the log store.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/datanode/src/datanode.rs [hunk 1]
```diff
@@ -490,6 +469,63 @@ impl DatanodeBuilder {
...
+
+    /// Builds [RaftEngineLogStore].
+    async fn build_raft_engine_log_store(
+        data_home: &str,
+        config: &RaftEngineConfig,
+    ) -> Result<Arc<RaftEngineLogStore>> {
+        let data_home = normalize_dir(data_home);
+        let wal_dir = match &config.dir {
+            Some(dir) => dir.clone(),
+            None => format!("{}{WAL_DIR}", data_home),
```
Comment (MichaelScofield) [inline src/datanode/src/datanode.rs:504]:
```suggestion
            None => format!("{}{WAL_DIR}", normalize_dir(data_home)),
```

File: src/datanode/src/datanode.rs [hunk 2]
```diff
@@ -346,11 +345,22 @@ impl DatanodeBuilder {
...
+                // Augments region options with wal options.
+                // TODO(niebayes): fetch wal options map from region info.
+                // let wal_options_map = &table_value.region_info.wal_options_map;
+                let wal_options_map: HashMap<RegionNumber, EncodedWalOptions> = HashMap::default();
+                let wal_options = wal_options_map
+                    .get(&region_number)
+                    .cloned()
+                    .unwrap_or_default();
+                let mut region_options = table_value.region_info.region_options.clone();
+                region_options.extend(wal_options);
```
Comment (MichaelScofield) [inline src/datanode/src/datanode.rs:357]: Can we pass in `wal_options` directly instead of put it in `region_options`? `extend` might implicitly clone the vec.

File: src/mito2/src/wal.rs [hunk 3]
```diff
@@ -126,8 +142,19 @@ impl<S: LogStore> WalWriter<S> {
...
         entry_id: EntryId,
         wal_entry: &WalEntry,
+        wal_options: &WalOptions,
     ) -> Result<()> {
-        let namespace = self.store.namespace(region_id.into());
+        // Gets or inserts with a newly built namespace.
+        let namespace = self
+            .namespaces
+            .entry(region_id)
+            .or_insert(try_build_wal_namespace(
```
Comment (MichaelScofield) [inline src/mito2/src/wal.rs:151]:
```suggestion
            .or_insert_with(|| try_build_wal_namespace(
```

File: src/store-api/src/logstore.rs [hunk 4]
```diff
@@ -59,19 +66,37 @@ pub trait LogStore: Send + Sync + 'static + std::fmt::Debug {
...
+        -> Self::Entry;
 
     /// Create a namespace of the associate Namespace type
     // TODO(sunng87): confusion with `create_namespace`
-    fn namespace(&self, id: namespace::Id) -> Self::Namespace;
+    fn namespace(
+        &self,
+        ns_id: NamespaceId,
+        wal_options: &WalOptions,
+    ) -> Result<Self::Namespace, Self::Error>;
```
Comment (MichaelScofield) [inline src/store-api/src/logstore.rs:78]: When does the `namespace` of kafka impl return an err?

File: src/store-api/src/logstore/entry.rs [hunk 5]
```diff
@@ -16,9 +16,12 @@ use common_error::ext::ErrorExt;
 
 use crate::logstore::namespace::Namespace;
 
+/// An entry's logical id, allocated by log store users.
+pub type Id = u64;
+/// An entry's physical offset in the underlying log store.
 pub type Offset = usize;
+// TODO(niebayes): consider removing `Epoch`.
```
Comment (v0y4g3r) [inline src/store-api/src/logstore/entry.rs:23]: You can safely remove it.

File: config/datanode.example.toml [hunk 6]
```diff
@@ -29,16 +29,36 @@ connect_timeout = "1s"
 # `TCP_NODELAY` option for accepted connections, true by default.
 tcp_nodelay = true
 
-# WAL options, see `standalone.example.toml`.
+# WAL options.
+# Uncomments all other wal options except the one you chose.
 [wal]
 # WAL data directory
+provider = "raft-engine"
```
Comment (evenyag) [inline config/datanode.example.toml:36]: Maybe we should use snake_case. @WenyXu Which do you prefer?
https://github.com/GreptimeTeam/greptimedb/issues/2936

Comment (WenyXu) [reply config/datanode.example.toml:36]: Yes, I vote for the `snake_case`,  too.

File: src/datanode/src/datanode.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (evenyag) [reply src/datanode/src/datanode.rs:357]: Is it possible to store a json string instead of a hashmap for wal options? The EncodedHashMap is hard to deserialize.

Another approach is just to store the enum in the map.

[the remark this answers is not in the document]
Comment (niebayes) [reply src/datanode/src/datanode.rs:357]: A `WalOptions` is simply a `String` now. See codes in another PR:
https://github.com/GreptimeTeam/greptimedb/blob/b2ec3d1c7126f3a98e8fd4f13764d9cf5c4f33d3/src/common/meta/src/wal.rs#L106-L126

Will update the deserialization for this PR soon.

File: config/datanode.example.toml (hunk 6 above)
Comment (niebayes) [reply config/datanode.example.toml:36]: Changed to `snake_case`.

File: config/datanode.example.toml [hunk 7]
```diff
@@ -29,16 +29,36 @@ connect_timeout = "1s"
...
+
+# Raft-engine wal options, see `standalone.example.toml`
```
[the remark this answers is not in the document]
Comment (niebayes) [reply config/datanode.example.toml:38]: Currently, users is expected to choose the wal through the provider field. When a wal provider is chose, the user should comment out all other wal config except ones that corresponds to the chosen one.

File: config/standalone.example.toml [hunk 8]
```diff
@@ -82,6 +82,13 @@ enable = true
 
 # WAL options.
 [wal]
+# Available wal providers:
+# - "RaftEngine" (default)
+# - "Kafka"
+provider = "raft-engine"
+
+# There's no kafka wal config for standalone mode.
```
[the remark this answers is not in the document]
Comment (niebayes) [reply config/standalone.example.toml:90]: Should the kafka be a cluster or a single broker in standalone mode? 
If the kafka broker can be embedded into the system as a component, we could support kafka wal in standalone mode in the future.

File: src/cmd/src/datanode.rs [hunk 9]
```diff
@@ -167,7 +168,13 @@ impl StartCommand {
         }
 
         if let Some(wal_dir) = &self.wal_dir {
-            opts.wal.dir = Some(wal_dir.clone());
+            // `wal_dir` only affects raft-engine config.
+            match &mut opts.wal {
+                WalConfig::RaftEngine(raft_engine_config) => {
+                    raft_engine_config.dir.replace(wal_dir.clone());
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/cmd/src/datanode.rs:174]: Added.

[the remark this answers is not in the document]
Comment (niebayes) [reply src/cmd/src/datanode.rs:174]:
```rust

        if let Some(wal_dir) = &self.wal_dir {
            // `wal_dir` only affects raft-engine config.
            match &mut opts.wal {
                WalConfig::RaftEngine(raft_engine_config) => {
                    if raft_engine_config
                        .dir
                        .as_ref()
                        .is_some_and(|original_dir| original_dir == wal_dir)
                    {
                        info!("The wal dir of raft-engine is altered to {wal_dir}");
                    }
                    raft_engine_config.dir.replace(wal_dir.clone());
                }
                WalConfig::Kafka(_) => {}
            }
        }
```

File: src/datanode/src/datanode.rs (hunk 1 above)
Comment (niebayes) [reply src/datanode/src/datanode.rs:504]: The `datanode` was already normalized. See line 501.

File: src/store-api/src/logstore.rs (hunk 4 above)
Comment (niebayes) [reply src/store-api/src/logstore.rs:78]: The interface is changed to not return a result since there're no cases that building namespace would incur an error.

File: config/datanode.example.toml [hunk 10]
```diff
@@ -29,16 +29,36 @@ connect_timeout = "1s"
...
+topic_name_prefix = "greptimedb_wal_kafka_topic"
+# Number of partitions per topic.
+num_partitions = 1
+# The maximum log size an rskafka batch producer could buffer.
+max_batch_size = "4MB"
+# The linger duration of an rskafka batch producer.
+linger = "200ms"
+# The maximum amount of time (in milliseconds) to wait for Kafka records to be returned.
+max_wait_time = "100ms"
+
```
[the remark this answers is not in the document]
Comment (niebayes) [reply config/datanode.example.toml:61]: Those are commented out.

File: src/store-api/src/logstore/entry.rs (hunk 5 above)
Comment (niebayes) [reply src/store-api/src/logstore/entry.rs:23]: Removed.

File: src/common/config/src/wal/kafka.rs [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+pub struct KafkaOptions {
+    /// Kafka wal topic.
+    pub topic: Topic,
+}
+
+impl Default for KafkaOptions {
+    fn default() -> Self {
+        Self {
+            // To indicates a default deserialized topic is invalid.
+            topic: "invalid_topic".to_string(),
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/common/config/src/wal/kafka.rs:79]: Removed the default implementation.

File: src/mito2/src/wal.rs (hunk 3 above)
Comment (niebayes) [reply src/mito2/src/wal.rs:151]: Changed.

File: config/datanode.example.toml [hunk 12]
```diff
@@ -29,16 +29,38 @@ connect_timeout = "1s"
...
 purge_threshold = "4GB"
 purge_interval = "10m"
 read_batch_size = 128
 sync_write = false
 
+# Kafka wal options.
+# The broker endpoints of the Kafka cluster. ["127.0.0.1:9090"] by default.
+# broker_endpoints = ["127.0.0.1:9090"]
+# Number of topics shall be created beforehand.
+# num_topics = 64
```
Comment (killme2008) [inline config/datanode.example.toml:52]: Why the datanode needs `num_topics`, `num_partitions` and `topic_name_prefix`? They are used in metasrv when allocating the region WAL, but it looks unnecessary for datanode.

File: config/standalone.example.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (killme2008) [reply config/standalone.example.toml:90]: The kafka can be a cluster or a single node, it doesn't matter.

File: src/common/config/src/wal/kafka.rs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    fn default() -> Self {
+        Self {
+            broker_endpoints: vec!["127.0.0.1:9090".to_string()],
+            num_topics: 64,
+            topic_name_prefix: TOPIC_NAME_PREFIX.to_string(),
+            num_partitions: 1,
+            compression: RsKafkaCompression::NoCompression,
+            max_batch_size: ReadableSize::mb(4),       // 4MB.
+            linger: Duration::from_millis(200),        // 200ms.
+            max_wait_time: Duration::from_millis(100), // 100ms.
```
Comment (killme2008) [inline src/common/config/src/wal/kafka.rs:62]: The code describes itself.
```suggestion
            max_batch_size: ReadableSize::mb(4),
            linger: Duration::from_millis(200),
            max_wait_time: Duration::from_millis(100),
```

File: src/common/config/src/wal/raft_engine.rs [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+// limitations under the License.
+
+use std::time::Duration;
+
+use common_base::readable_size::ReadableSize;
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(default)]
+pub struct RaftEngineConfig {
```
Comment (killme2008) [inline src/common/config/src/wal/raft_engine.rs:22]: Can we add a doc for this struct? I knew it didn't have before.

File: src/common/config/src/wal/raft_engine.rs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    // whether to sync log file after every write
+    pub sync_write: bool,
+}
+
+impl Default for RaftEngineConfig {
+    fn default() -> Self {
+        Self {
+            dir: None,
+            file_size: ReadableSize::mb(256), // log file size 256MB
+            purge_threshold: ReadableSize::gb(4), // purge threshold 4GB
```
Comment (killme2008) [inline src/common/config/src/wal/raft_engine.rs:43]:
```suggestion
            file_size: ReadableSize::mb(256),
            purge_threshold: ReadableSize::gb(4),
```

File: src/datanode/src/datanode.rs [hunk 16]
```diff
@@ -347,11 +346,22 @@ impl DatanodeBuilder {
         while let Some(table_value) = table_values.next().await {
             let table_value = table_value.context(GetMetadataSnafu)?;
             for region_number in table_value.regions {
+                // Augments region options with wal options.
+                // TODO(niebayes): fetch wal options map from region info.
+                // let wal_options_map = &table_value.region_info.wal_options_map;
```
Comment (killme2008) [inline src/datanode/src/datanode.rs:351]: Do we need to wait for another PR to fix this `TODO`?

File: src/datanode/src/datanode.rs [hunk 17]
```diff
@@ -493,6 +473,63 @@ impl DatanodeBuilder {
         Ok(engines)
     }
 
+    /// Builds [MitoEngine] according to options.
+    async fn build_mito_engine(
+        opts: &DatanodeOptions,
+        object_store_manager: ObjectStoreManagerRef,
+        config: MitoConfig,
+    ) -> Result<MitoEngine> {
+        // TODO(niebayes): validate wal config, for e.g. providing a remote wal for standalone mode is not allowed.
```
Comment (killme2008) [inline src/datanode/src/datanode.rs:482]: Looks like it's allowed, so we must modify the comment. But I think we should do the other validations right now, don't make it to be a `TODO`.

File: src/log-store/src/kafka.rs [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+pub mod log_store;
+
+use store_api::logstore::entry::{Entry, Id as EntryId};
+use store_api::logstore::namespace::Namespace;
+
+use crate::error::Error;
+
+type Topic = String;
+
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
```
Comment (killme2008) [inline src/log-store/src/kafka.rs:24]: I don't want to repeat myself. I think all the pub structures must have a doc to describe them.

File: src/mito2/src/region/options.rs [hunk 19]
```diff
@@ -52,11 +55,13 @@ impl TryFrom<&HashMap<String, String>> for RegionOptions {
         let options: RegionOptionsWithoutEnum =
             serde_json::from_str(&json).context(JsonOptionsSnafu)?;
         let compaction: CompactionOptions = serde_json::from_str(&json).unwrap_or_default();
+        let wal_options: WalOptions = serde_json::from_str(&json).unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (killme2008) [reply src/mito2/src/region/options.rs:58]: I wonder if the serialization is compatible with old versions, or it's a breaking change.

File: src/mito2/src/wal.rs [hunk 20]
```diff
@@ -60,13 +62,19 @@ impl<S: LogStore> Wal<S> {
             store: self.store.clone(),
             entries: Vec::new(),
             entry_encode_buf: Vec::new(),
+            namespaces: HashMap::new(),
         }
     }
 
     /// Scan entries of specific region starting from `start_id` (inclusive).
-    pub fn scan(&self, region_id: RegionId, start_id: EntryId) -> Result<WalEntryStream> {
+    pub fn scan<'a>(
```
Comment (killme2008) [inline src/mito2/src/wal.rs:70]: Question: I am not sure if we can remove the `'a` here. Looks like it's not necessary.

File: src/datanode/src/datanode.rs (hunk 16 above)
Comment (niebayes) [reply src/datanode/src/datanode.rs:351]: Since the meta srv skeleton PR was merged, this TODO is removed.

File: src/datanode/src/datanode.rs (hunk 2 above)
Comment (niebayes) [reply src/datanode/src/datanode.rs:358]: The `wal_options` is now encoded as a string and hence no need to extend.

File: config/standalone.example.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply config/standalone.example.toml:90]: Standalone mode now supports allocating kafka wal topics when handling create table requests.

File: src/cmd/src/datanode.rs [hunk 21]
```diff
@@ -167,7 +168,13 @@ impl StartCommand {
...
+                }
+                WalConfig::Kafka(_) => {}
+            }
         }
```
[the remark this answers is not in the document]
Comment (niebayes) [reply src/cmd/src/datanode.rs]: Chaining let is a unstable feature. See: https://github.com/rust-lang/rust/issues/53667

File: src/mito2/src/region/options.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/mito2/src/region/options.rs:58]: @killme2008 The deserialization scheme for `wal_options` is changed and it's compatible with the legacy RegionOptions. No need to add a compatible test since the existing tests already checked the compatibility.

[the remark this answers is not in the document]
Comment (niebayes) [reply src/mito2/src/region/options.rs:58]: @v0y4g3r  The `try_from` implemented is changed. See: 
```rust
    fn try_from(options_map: &HashMap<String, String>) -> Result<Self> {
        let value = options_map_to_value(options_map);
        let json = serde_json::to_string(&value).context(JsonOptionsSnafu)?;

        // #[serde(flatten)] doesn't work with #[serde(default)] so we need to parse
        // each field manually instead of using #[serde(flatten)] for `compaction`.
        // See https://github.com/serde-rs/serde/issues/1626
        let options: RegionOptionsWithoutEnum =
            serde_json::from_str(&json).context(JsonOptionsSnafu)?;
        let compaction: CompactionOptions = serde_json::from_str(&json).unwrap_or_default();

        // Tries to decode the wal options from the map or sets to the default if there's none wal options in the map.
        let wal_options = options_map.get(WAL_OPTIONS_KEY).map_or_else(
            || Ok(WalOptions::default()),
            |encoded_wal_options| {
                serde_json::from_str(&encoded_wal_options).context(JsonOptionsSnafu)
            },
        )?;

        Ok(RegionOptions {
            ttl: options.ttl,
            compaction,
            storage: options.storage,
            wal_options,
        })
    }
```

File: config/datanode.example.toml (hunk 12 above)
Comment (niebayes) [reply config/datanode.example.toml:52]: Yes. Some configs are not necessary. Will remove unnecessary ones when implementing kafka log store.

File: src/log-store/src/kafka.rs (hunk 18 above)
Comment (niebayes) [reply src/log-store/src/kafka.rs:24]: Added.

File: src/common/config/src/wal/raft_engine.rs (hunk 14 above)
Comment (niebayes) [reply src/common/config/src/wal/raft_engine.rs:23]: Added.

File: src/datanode/src/datanode.rs (hunk 17 above)
Comment (niebayes) [reply src/datanode/src/datanode.rs:482]: The TODO is removed since we allow standalone mode to also use remote wal.
The type system already applies enough constraints on the wal config and currently there's no need to do extra validations.

File: src/cmd/src/datanode.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/cmd/src/datanode.rs]: > Chaining let is a unstable feature. See: [rust-lang/rust#53667](https://github.com/rust-lang/rust/issues/53667)

We've already introduced this feature multiple times, for example:https://github.com/GreptimeTeam/greptimedb/blob/14e6998d4160b9f6d1fe4fd0a9cb27bb13a92c9f/src/datatypes/src/lib.rs#L15

File: src/mito2/src/wal.rs (hunk 20 above)
Comment (niebayes) [reply src/mito2/src/wal.rs:70]: If without the lifetime, the compiler reports an error: 
```rust
error: lifetime may not live long enough
  --> src/mito2/src/wal.rs:96:9
   |
71 |         & self,
   |         - let's call the lifetime of this reference `'2`
...
74 |         wal_options: &WalOptions,
   |                      - let's call the lifetime of this reference `'1`
...
96 |         Ok(Box::pin(stream))
   |         ^^^^^^^^^^^^^^^^^^^^ method was supposed to return data with lifetime `'2` but it is returning data with lifetime `'1`
   |
help: consider introducing a named lifetime parameter and update trait if needed
   |
70 ~     pub fn scan<'a>(
71 ~         &'a  self,
72 |         region_id: RegionId,
73 |         start_id: EntryId,
74 ~         wal_options: &'a WalOptions,
   |
```

File: src/cmd/src/datanode.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (niebayes) [reply src/cmd/src/datanode.rs]: Updated.

**Review by killme2008 — APPROVED** [review]
LGTM

**Review by v0y4g3r — APPROVED** [review]
LGTM
